fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -428,6 +436,37 @@ public RecordMetaDataBuilder setRecords(@Nonnull Descriptors.FileDescriptor file
         return this;
     }
 
+    /**
+     * Update the records descriptor of the record meta-data.
+     *
+     * <p>
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
+     * By contract, the extension options will not be processed for the old record types.
+     * </p>
+     *
+     * @param recordsDescriptor the new record descriptor
+     */
+    public void updateRecords(Descriptors.FileDescriptor recordsDescriptor) {
+        updateRecords(recordsDescriptor, true);
+    }
+
+    /**
+     * Update the record descriptor of the record meta-data.
+     *
+     * <p>
+     * This involves adding new record types and updating descriptors for the existing record types and union fields.
+     * By contract, the extension options will not be processed for the old record types.
+     * </p>
+     *
+     * @param recordsDescriptor the new record descriptor
+     * @param processExtensionOptions whether to add primary keys and indexes using the extensions in the protobuf (only for the new record types)
+     */
+    public void updateRecords(Descriptors.FileDescriptor recordsDescriptor, boolean processExtensionOptions) {
+        this.recordsDescriptor = recordsDescriptor;
+        unionDescriptor = updateRecordTypes(this.recordsDescriptor, processExtensionOptions);
+        processSchemaOptions(processExtensionOptions);
```

Comment:
Yeah. The reason not to process the schema options is that there are already ways of setting them in the builder (i.e., `setSplitLongRecords` and `setStoreRecordVersions`), and I think we want to reserve this method for doing things that *cannot* be done through other mechanisms (which is to say, updating the actual message descriptors of the various record types) to avoid having the user accidentally mutate other things.

I think the one exception to that rule would be adding indexes and primary key information from the new record types based on those types extension information...but even then, I'm somewhat lukewarm to that idea.